File: 백준/Gold/7579. 앱/앱.java
/**
 * Author: nowalex322, Kim HyeonJae
 */

import java.io.*;
import java.util.*;

public class Main {
    static BufferedReader br;
    static BufferedWriter bw;
    static StringTokenizer st;
	static int N, M;
	static int[] memory, cost;
    public static void main(String[] args) throws Exception {
        new Main().solution();
    }

    public void solution() throws Exception {
        br = new BufferedReader(new InputStreamReader(System.in));
        //br = new BufferedReader(new InputStreamReader(new FileInputStream("src/main/java/BOJ_7579_앱/input.txt")));
        bw = new BufferedWriter(new OutputStreamWriter(System.out));

		st = new StringTokenizer(br.readLine());
		N = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());

		memory = new int[N];
		st = new StringTokenizer(br.readLine());
		for(int i=0; i<N; i++) {
			memory[i] = Integer.parseInt(st.nextToken());
		}

		cost = new int[N];
		int maxCost = 0;
		st = new StringTokenizer(br.readLine());
		for(int i=0; i<N; i++) {
			cost[i] = Integer.parseInt(st.nextToken());
			maxCost += cost[i];
		}

		int[] dp = new int[maxCost + 1]; // dp[i]= i비용으로 확보가능한 최대 메모리
		for(int i=0; i<N; i++) {
			for(int j=maxCost; j>=cost[i]; j--){
				dp[j] = Math.max(dp[j], dp[j-cost[i]] + memory[i]);
			}
		}

		for(int i=0; i<=maxCost; i++) {
			if(dp[i] >= M) {
				System.out.println(i);
				break;
			}
		}


        bw.flush();
        bw.close();
        br.close();
    }
}

File: 백준/Gold/7579. 앱/앱.kt
/**
 * Author: nowalex322, Kim HyeonJae
 */

import java.util.*;
import java.io.*;

private lateinit var br: BufferedReader
private lateinit var bw: BufferedWriter
private var N=0
private var M=0
private lateinit var memory: IntArray
private lateinit var cost: IntArray

fun main(){
    br = BufferedReader(InputStreamReader(System.`in`))
    bw = BufferedWriter(OutputStreamWriter(System.out))

    val (n, m) = br.readLine().split(" ").map { it.toInt() }
    N = n
    M = m

    memory = br.readLine().split(" ").map { it.toInt() }.toIntArray()
    cost = br.readLine().split(" ").map { it.toInt() }.toIntArray()

    val maxCost = cost.sum()

    val dp = IntArray(maxCost+1)

    for (i in 0 until N) {
        for (j in maxCost downTo cost[i]) {
            dp[j] = maxOf(dp[j], dp[j - cost[i]] + memory[i])
        }
    }

    for (i in 0..maxCost) {
        if (dp[i] >= M) {
            println(i)
            break
        }
    }

}